Add v2 update docs

pirj opened the pull request:
Better viewed as https://github.com/rubocop-hq/rubocop-rspec/blob/add-v2-migration-docs/docs/modules/ROOT/pages/upgrade_to_version_2.adoc

### TODO

- [x] adjust docs related to the changes from #1063
  - [x] that `Include` should be used instead of `Patterns`
  - [x] note that `Include`'s merge mode is set to override by default
- [x] adjust docs related to the final changes to API from #956
- - [x] syntax changes
- - [x] mention the change of ExampleGroups::GROUPS to ExampleGroups#regular and from Examples::EXAMPLES to Examples#regular https://github.com/rubocop-hq/rubocop-rspec/pull/956/files#r516238107
- - [x]       # call `super` if you define on_new_investigation in a cop
- - [x] do not use `let(:cop) { described_class.new }` in custom cops' specs, as this will break specs (config not loaded)
- - [ ] add a section on how to configure DSL aliases


---

Before submitting the PR make sure the following are checked:

* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [-] Added tests.
* [x] Updated documentation.
* [-] Added an entry to the `CHANGELOG.md` if the new code introduces user-observable changes.
* [x] The build (`bundle exec rake`) passes (be sure to run this locally, since it may produce updated documentation that you will need to commit).

pirj commented:
@sl4vr appreciate if you could take a look.

pirj commented:
@bquorning Green.
Besides all, this includes information on how to migrate to 2.0 given we decide to merge #956 in its current form.

bquorning commented on docs/modules/ROOT/pages/upgrade_to_version_2.adoc:
```diff
@@ -0,0 +1,166 @@
+Due to significant API changes custom cops will break.
+Here is the summary of the changes:
+
+1) The base class for cops is now `RuboCop::Cop::RSpec::Base` instead of `RuboCop::Cop::RSpec::Cop`.
```
I think this line should start with `1.`
https://asciidoctor.org/docs/asciidoc-writers-guide/#ordering-the-things

pirj replied:
Parrotted it from here https://github.com/rubocop-hq/rubocop/blame/master/docs/modules/ROOT/pages/v1_upgrade_notes.adoc#L10 that is quite good on the eye https://docs.rubocop.org/rubocop/v1_upgrade_notes.html

bquorning replied:
> quite good on the eye

Looks like I’ll have to open a pull request to fix RuboCop’s documentation too. 😄 

IMO this

<img width="403" alt="Screenshot 2020-10-22 at 22 27 09" src="https://user-images.githubusercontent.com/22333/96926409-cfa0f380-14b5-11eb-8df3-738c212a8e02.png">

looks more pleasing than this

<img width="403" alt="Screenshot 2020-10-22 at 22 22 31" src="https://user-images.githubusercontent.com/22333/96926424-d465a780-14b5-11eb-99f6-36a45b3f841e.png">

(note especially the indentation of the 2nd point)

bquorning replied:
Again, sorry for my nit-pick comments.

pirj replied:
Agree, AsciiDoc's list looks better.

> Looks like I’ll have to open a pull request to fix RuboCop’s documentation too. 

:laughing: :shrug:

pirj replied:
> sorry for my nit-pick comments

No worries, those comments are all to the point. :+1:

bquorning commented on docs/modules/ROOT/pages/upgrade_to_version_2.adoc:
```diff
@@ -0,0 +1,166 @@
+
+== Custom Cop Update Guide
+
+Due to significant API changes custom cops will break.
```
```suggestion
Due to significant API changes, custom cops may break.
```

pirj replied:
With a 95% chance, I'd say :D

bquorning commented on docs/modules/ROOT/pages/upgrade_to_version_2.adoc:
```diff
@@ -0,0 +1,166 @@
+
+2) `TopLevelDescribe` is replaced with a more generic `TopLevelGroup`.
+
+3) `RuboCop::RSpec::Language` has been completely rewritten to support dynamic RSpec DSL aliases and negated matchers to fully support third-party libraries e.g. RSpec Rails, Pundit, Action Policy and many others.
```
```suggestion
3) `RuboCop::RSpec::Language` has been completely rewritten to support dynamic RSpec DSL aliases and negated matchers to fully support third-party libraries such as RSpec Rails, Pundit, Action Policy and many others.
```

bquorning commented on docs/modules/ROOT/pages/upgrade_to_version_2.adoc:
```diff
@@ -0,0 +1,166 @@
+
+=== Change the `Language` Module Usages
+
+The `RuboCop::RSpec::Language` is completely different now.
```
I still have to get used to this fact 😅

pirj replied:
Me too, but we have plenty of reference examples luckily :D

bquorning commented on docs/modules/ROOT/pages/upgrade_to_version_2.adoc:
```diff
@@ -0,0 +1,166 @@
+
+# After
+def_node_search :examples?,
+                send_pattern(
+                  '{#rspec(:Includes, :Example) #rspec(:Examples)}')
```
I think it would be better if we can keep this on one line.
```suggestion
                send_pattern('{#rspec(:Includes, :Example) #rspec(:Examples)}')
```
BTW it is _exactly_ 80 characters wide 😅

bquorning commented on docs/modules/ROOT/pages/upgrade_to_version_2.adoc:
```diff
@@ -0,0 +1,166 @@
+
+=== Conform with RuboCop API Changes
+
+The parent project, RuboCop, have API changes.
```
```suggestion
The parent project, RuboCop, has API changes.
```

bquorning commented on docs/modules/ROOT/pages/upgrade_to_version_2.adoc:
```diff
@@ -0,0 +1,166 @@
+=== Conform with RuboCop API Changes
+
+The parent project, RuboCop, have API changes.
+They won't result in cop breakages, however, it's recommended to update cops to use new APIs.
```
```suggestion
While they won’t result in cop breakages, it is recommended to update cops to use new API’s.
```

bquorning commented on docs/modules/ROOT/pages/upgrade_to_version_2.adoc:
```diff
@@ -0,0 +1,166 @@
+
+The parent project, RuboCop, have API changes.
+They won't result in cop breakages, however, it's recommended to update cops to use new APIs.
+Follow the https://docs.rubocop.org/rubocop/v1_upgrade_notes[RuboCop's core v1 update guide] to adjust custom cops' RuboCop Autocorrect API usage.
```
```suggestion
Follow the https://docs.rubocop.org/rubocop/v1_upgrade_notes[RuboCop v1 update guide] to adjust custom cops’ use of RuboCop’s auto-correction API.
```

bquorning commented on docs/modules/ROOT/pages/upgrade_to_version_2.adoc:
```diff
@@ -0,0 +1,166 @@
+
+ - `RSpec/InvalidPredicateMatcher` cop is removed
+ - `CustomIncludeMethods` configuration option for `RSpec/EmptyExampleGroup` is removed
+ - cop departments are nested for cops with a department that doesn't match the extension name (`Capybara`, `FactoryBot`, `Rails`)
```
```suggestion
 - cop departments are nested for cops with a department that doesn’t match the extension name (`Capybara`, `FactoryBot`, `Rails`)
```

bquorning commented on docs/modules/ROOT/pages/upgrade_to_version_2.adoc:
```diff
@@ -0,0 +1,166 @@
+
+RuboCop extensions had cops with clashing names and departments, e.g. both `rspec-rails` and `rubocop-rspec` had `Rails::HttpStatus` cops.
+To avoid issues, e.g. inability to disable just one of the cops, each extension now has its own uber-department.
+Expectedly, RuboCop RSpec's uber-department name is `RSpec`.
```
```suggestion
Expectedly, RuboCop RSpec’s uber-department name is `RSpec`.
```

pirj replied:
I'm still not sure if "uber-department" is a correct term, but couldn't think of a better one :D

bquorning commented on docs/modules/ROOT/pages/upgrade_to_version_2.adoc:
```diff
@@ -0,0 +1,166 @@
+RuboCop extensions had cops with clashing names and departments, e.g. both `rspec-rails` and `rubocop-rspec` had `Rails::HttpStatus` cops.
+To avoid issues, e.g. inability to disable just one of the cops, each extension now has its own uber-department.
+Expectedly, RuboCop RSpec's uber-department name is `RSpec`.
+Changes are only applied to cops that don't already have the department set to `RSpec`, i.e. `Capybara`, `FactoryBot` and `Rails`.
```
```suggestion
Changes are only applied to cops that don’t already have the department set to `RSpec`, i.e. `Capybara`, `FactoryBot` and `Rails`.
```

bquorning commented on docs/modules/ROOT/pages/upgrade_to_version_2.adoc:
```diff
@@ -0,0 +1,166 @@
+`TopLevelDescribe` was incomplete, had poor performance and did not distinguish between example groups and shared example groups.
+
+`TopLevelGroup` provides a similar interface, but instead of a single `on_top_level_describe` hook there are two, `on_top_level_example_group` and `on_top_level_group`.
+There's no need yet for `on_top_level_shared_group` for RuboCop core cops, but if your custom cop needs such a hook, please feel free to send a pull request.
```
```suggestion
There’s no need yet for `on_top_level_shared_group` for RuboCop core cops, but if your custom cop needs such a hook, please feel free to send a pull request.
```

bquorning reviewed (commented):
Sorry about the many nitpick comments here.

I believe this PR should be the last one to be merged into `release-2.0`, as it may still change depending on the other PRs getting merged.

pirj commented:
Thanks for the proofreading, @bquorning !

> I believe this PR should be the last one to be merged into release-2.0

Agree :+1:

pirj commented:
All review notes applied :heavy_check_mark:

pirj commented:
@bquorning Ready for the review. There are changes since the last review, the list is in TODO section of the ticket description.
I've rearranged all the commits.

sl4vr commented on docs/modules/ROOT/pages/upgrade_to_version_2.adoc:
```diff
@@ -0,0 +1,273 @@
+ExampleGroups::GROUPS
+
+# After
+ExampleGroups#regular
```
```suggestion
ExampleGroups.regular
```

sl4vr commented on docs/modules/ROOT/pages/upgrade_to_version_2.adoc:
```diff
@@ -0,0 +1,273 @@
+Examples::EXAMPLES
+
+# After
+Examples#regular
```
```suggestion
Examples.regular
```

sl4vr reviewed (commented):
Just found a few inaccuracies.

pirj commented:
Eagle eye, @sl4vr, thanks a lot!

bquorning commented on docs/modules/ROOT/pages/usage.adoc:
```diff
@@ -1,19 +1,80 @@
+----
+# .rubocop.yml
+
+require:
+ - rubocop-rspec
+ - test-prof
```
```suggestion
require:
  - rubocop-rspec
  - test-prof
```

bquorning commented on docs/modules/ROOT/pages/usage.adoc:
```diff
@@ -1,19 +1,80 @@
+or, if you are using several extensions:
+
+----
+require:
+ - rubocop-rspec
+ - rubocop-performance
```
```suggestion
require:
  - rubocop-rspec
  - rubocop-performance
```

bquorning commented on docs/modules/ROOT/pages/usage.adoc:
```diff
@@ -1,19 +1,80 @@
 
+=== RSpec DSL configuration
+
+In case you https://github.com/rspec/rspec-core/blob/b0d0843a285693c64cdbe0c85726db155b46047e/lib/rspec/core/configuration.rb#L1122[define aliases for RSpec DSL], i.e. examples, example groups, hooks, or include example statements, you need to properly configure it so those elements are properly detected by RuboCop-RSpec.
```
```suggestion
In case you https://github.com/rspec/rspec-core/blob/b0d0843a285693c64cdbe0c85726db155b46047e/lib/rspec/core/configuration.rb#L1122[define aliases for RSpec DSL], i.e. examples, example groups, hooks, or include example statements, you need to configure it so those elements are properly detected by RuboCop RSpec.
```
You had `properly` twice in the same sentence there.

bquorning commented on docs/modules/ROOT/pages/upgrade_to_version_2.adoc:
```diff
@@ -0,0 +1,273 @@
+
+=== Use the RuboCop standard `Include` option to filter inspected files
+
+`Patterns` was a RuboCop-RSpec-specific option, and RuboCop has a standard replacement.
```
```suggestion
`Patterns` was a RuboCop RSpec-specific option, and RuboCop has a standard replacement.
```

bquorning commented on docs/modules/ROOT/pages/usage.adoc:
```diff
@@ -1,19 +1,80 @@
+
+NOTE: the default merge mode is to inherit, so you won't remove any of the default settings.
+
+RuboCop-RSpec's https://github.com/rubocop-hq/rubocop-rspec/blob/a43424527c09fae2e6ddb133f4b2988f6c46bb2e/config/default.yml#L6[default configuration] is a good source of information on what can be configured.
```
```suggestion
RuboCop RSpec's https://github.com/rubocop-hq/rubocop-rspec/blob/a43424527c09fae2e6ddb133f4b2988f6c46bb2e/config/default.yml#L6[default configuration] is a good source of information on what can be configured.
```

bquorning commented on docs/modules/ROOT/pages/usage.adoc:
```diff
@@ -1,19 +1,80 @@
+[source,ruby]
+----
+# spec/spec_helper.rb
+alias_example_group_to
```
I think this line is not needed.

bquorning commented on docs/modules/ROOT/pages/usage.adoc:
```diff
@@ -1,19 +1,80 @@
+        - detail
+----
+
+Some libraries extensively define RSpec DSL aliases (e.g. Pundit, ActionPolicy) or augment existing elements providing the same semantics (e.g. `let_it_be` from `test-prof`)
```
```suggestion
Some libraries extensively define RSpec DSL aliases (e.g. Pundit, Action Policy) or augment existing elements providing the same semantics (e.g. `let_it_be` from `test-prof`).
```

bquorning commented on docs/modules/ROOT/pages/usage.adoc:
```diff
@@ -53,3 +114,5 @@ RSpec:
     - '**/*_test.rb'
 ----
+
+NOTE: Please keep in mind that `Include`'s merge mode is set to override the default settings, so if you intend to add a path, while keeping the default paths, you should include the default include paths in your configuration.
```
```suggestion
NOTE: Please keep in mind that `Include`'s merge mode is set to override the default settings, so if you intend to add a path while keeping the default paths, you should include the default `Include` paths in your configuration.
```

bquorning commented on docs/modules/ROOT/pages/upgrade_to_version_2.adoc:
```diff
@@ -0,0 +1,273 @@
+ - `CustomIncludeMethods` configuration option for `RSpec/EmptyExampleGroup` is removed
+ - cop departments are nested for cops with a department that doesn’t match the extension name (`Capybara`, `FactoryBot`, `Rails`)
+ - `AllCops/RSpec/Patterns`/`AllCops/FactoryBot/Patterns` options are removed
+ - Calling `super` from `on_new_investigation` defined in a cop is mandatory now
```
```suggestion
 - Calling `super` from `#on_new_investigation` defined in a cop is mandatory now
```

bquorning commented on docs/modules/ROOT/pages/upgrade_to_version_2.adoc:
```diff
@@ -0,0 +1,273 @@
+ - cop departments are nested for cops with a department that doesn’t match the extension name (`Capybara`, `FactoryBot`, `Rails`)
+ - `AllCops/RSpec/Patterns`/`AllCops/FactoryBot/Patterns` options are removed
+ - Calling `super` from `on_new_investigation` defined in a cop is mandatory now
+ - Avoid explicit definition of `cop` as `described_class.new`
```
I don’t understand this sentence. Could you explain a bit more, please?

bquorning commented on docs/modules/ROOT/pages/upgrade_to_version_2.adoc:
```diff
@@ -0,0 +1,273 @@
+# .rubocop.yml
+
+# Before
+AllCops:
+  RSpec/FactoryBot:
+    Patterns:
+    - spec/factories/**/*.rb
+    - property/factories/**/*.rb
```
```suggestion
AllCops:
  RSpec/FactoryBot:
    Patterns:
      - spec/factories/**/*.rb
      - property/factories/**/*.rb
```

bquorning commented on docs/modules/ROOT/pages/upgrade_to_version_2.adoc:
```diff
@@ -0,0 +1,273 @@
+# .rubocop.yml
+
+# Before
+RSpec/EmptyExampleGroup:
+  CustomIncludeMethods:
+  - include_tests
+
+# After
+RSpec:
+  Language:
+    Includes:
+      Examples:
+      - include_tests
```
```suggestion
RSpec/EmptyExampleGroup:
  CustomIncludeMethods:
    - include_tests

# After
RSpec:
  Language:
    Includes:
      Examples:
        - include_tests
```

bquorning commented on docs/modules/ROOT/pages/upgrade_to_version_2.adoc:
```diff
@@ -0,0 +1,273 @@
+    - property/factories/**/*.rb
+----
+
+NOTE: Please keep in mind that `Include`'s merge mode is set to override the default settings, so if you intend to add a path, while keeping the default paths, you should include the default include paths in your configuration.
```
```suggestion
NOTE: Please keep in mind that `Include`'s merge mode is set to override the default settings, so if you intend to add a path while keeping the default paths, you should include the default `Include` paths in your configuration.
```

pirj replied:
"Include the include" 🤦

bquorning commented on docs/modules/ROOT/pages/upgrade_to_version_2.adoc:
```diff
@@ -0,0 +1,273 @@
+
+1. The base class for cops is now `RuboCop::Cop::RSpec::Base` instead of `RuboCop::Cop::RSpec::Cop`.
+
+2. `TopLevelDescribe` is replaced with a more generic `TopLevelGroup`.
```
```suggestion
2. The module `RuboCop::Cop::RSpec::TopLevelDescribe` is replaced with a more generic `RuboCop::Cop::RSpec::TopLevelGroup`.
```

pirj commented:
Suggestions applied, clarifications made. Thanks for the review, @bquorning !

pirj commented:
`edge-rubocop` is back to green! 💚
